plugins/trace_api_plugin/include/eosio/trace_api_plugin/store_provider.hpp

```diff
@@ -0,0 +1,186 @@
+#pragma once
+
+#include <ios>
+#include <fc/io/cfile.hpp>
+#include <boost/filesystem.hpp>
+#include <fc/variant.hpp>
+#include <eosio/trace_api_plugin/metadata_log.hpp>
+#include <eosio/trace_api_plugin/data_log.hpp>
+
+namespace eosio::trace_api_plugin {
+   using namespace boost::filesystem;
+
+   class path_does_not_exist : public std::runtime_error {
+   public:
+      explicit path_does_not_exist(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit path_does_not_exist(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class old_slice_version : public std::runtime_error {
+   public:
+      explicit old_slice_version(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit old_slice_version(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class incompatible_slice_files : public std::runtime_error {
+   public:
+      explicit incompatible_slice_files(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit incompatible_slice_files(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class malformed_slice_file : public std::runtime_error {
+   public:
+      explicit malformed_slice_file(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit malformed_slice_file(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class store_provider;
+
+
+   class slice_provider {
```

Comment:
This seems like an unused abstraction, was the intent to allow `slice_provider` to be swapped for some other implementation at some point?  

Even then it doesn't seem like a clean abstraction to me.  Either
- The `slice_provider` is abstracting *only* writes to/from the filesystem (though it seems like `store_provider` is tightly coupled to) in which case `slice_number()` seems out of place.  The division of data into buckets/slices is owned by the caller
- The `slice_provider` is abstracting the division of data *and* the filesystem in which case the caller should not need to be aware of the `slice_number` it should just push data down and let that next system divide it up and own the edge cases.  

My recommendation is to remove this layer of abstraction and roll it into `store_provider` as it seems 1:1 coupled to that class.

As you already have `vslice` for testing, I could see this becoming the non-mock implementation of a slice that is backed by `fc::cfile` but that would require that `store_provider` be completely abstracted from `fc::cfile`.  That seems like more work at this point.

Reply:
This is really for testing, if we pull it into store_provider, my preference would still be to have it as a inner class, but if these are private I'll need to add a test_friend class to allow testing.  I think putting a comment for slice_provider indicating that it is only intended to be used by store_provider and unit tests.  I will admit that having them both use *_provider will lead to confusion.  Would prefer to make that my next task (after this gets merged).